Side Panels Implementation

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

Follows up on previous work in #8160 which was merged for String Freeze.

`FullScreenSidePanel` is a component in `kolibri.coreVue.components` that provides responsive functionality and a default slot into which content must be added. This allows this backdrop-masked side bar to be used across Kolibri without coupling it to any particular data.

This PR covers two kinds of side panel content, accessible in three places by the user.

[Scenario 1 - A User is Viewing Content](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=328%3A25722)

| Actual Content | With Fake Data |
|-|-|
|![actual-irl-video-metadata](https://user-images.githubusercontent.com/6356129/134726937-3370e523-bccb-4c42-9299-60bddc5e62c7.png)|![metadata-with-all-the-fixins](https://user-images.githubusercontent.com/6356129/134726941-75892d89-882e-4bf0-b2a1-413d07c7e083.png)|



[Scenario 2 - A User is Browsing Content](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=55%3A8)

`SidePanelResourceMetadata` was initially the implementation covering one of the display styles. I opted to strip this component and base two others off of it. This resulted in my using `kolibri.utils.i18n.crossComponentTranslator` to get these strings into the two new components I've added covering the two scenarios:

1) `CurrentlyViewedResourceMetadata` covers the first scenario and is rendered in `LearnImmersiveLayout`
2) `BrowseResourceMetadata` covers the second scenario and is rendered in `LibraryPage` and `TopicsPage`

**Important notes**

- This focuses firstly on providing the styles and layout so that once we decide on how the metadata is structured, we can wire it all up to working layout.
- Ideally, the data passed to the `*ResourceMetadata` components will readily and easily include everything needed - so any state mapping can be done higher in the hierarchy. It would be super nice to have a bespoke data type - perhaps returned from a getter - that can be made readily available to components rendering the side panels.
- There is a lot of duplication here but that feels better to me than trying to work out some convoluted system for displaying the metadata in a `v-for` of some kind. I'm definitely open to suggestions on this approach or ways we might be able to improve.


## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

#8107

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

The best way to test this is locally in the code and by putting fake data in for each metadata slot if you want to see it.

I don't really think there is much to test aside from making sure the side bars are toggling correctly, have the correct metadata in place per the specs and are on the right track to being ready once we know the metadata in a concrete way.

Note, the "related resources" is missing. We can work through that after the fact, this is an aging PR and is stable enough to get in and we can build up on it in coming weeks.

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

nucleogenesis committed:
content metadata fullscreen side panel

nucleogenesis committed:
split SidePanelResourceMetadata into multiple, cxComponentT8n, add to lib and topics page

MisRob commented:
I'll skim through the code soon. Thank you for the summary, this approach sounds good to me.

Regarding TODOs, I am not sure if the following will be helpful in all cases, but if you'd like to you can check it out and see, or adjust as needed:

> Derive and display thumbnail (get the styles correct)

`ContentNodeThumbnail` and `QuizThumbnail` made for hybrid learning features [here](https://github.com/learningequality/kolibri/tree/develop/kolibri/plugins/learn/assets/src/views/thumbnails) might be helpful

> Figure out how / where to get Related resources & "location" in channel, where they link to, etc for BrowseResourceMetadata

Regarding related resources, @rtibbles recommended using our existing endpoint for now so I got them from [recommended getter](https://github.com/MisRob/kolibri/blob/0c3125b5a81c4b28ec3e8fc1e09e6be29690cc6f/kolibri/plugins/learn/assets/src/views/ContentPage.vue#L202) (however, this getter seems to be associated with the current topic tree) and fetched them with [this function](https://github.com/learningequality/kolibri/pull/8267/files#diff-deb71a7e53686b5ffe1c519dae8566920a62f425019dcf4416593c19947af6d1R23) for [the completion modal](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=2254%3A6) If these related resources are supposed to be the same as those on the modal, some of this logic might help.

MisRob commented:
I've checked code focusing on the high-level architecture of side panels and having two components, `CurrentlyViewedResourceMetadata` and `BrowseResourceMetadata`, seems to be simpler to me too. Even from designs, there are relatively lots of differences, so I consider this direction to be good.

 Regarding duplicates, I wouldn't be too worried about it. We can always extract and start sharing areas that intersect. Might be rather my opinion - I find it simpler than guessing how to generalize. I see two areas that we could improve, not necessarily in this PR though:

**Metadata**

I agree that using `v-for` would most likely end up very complex since we also need to format various metadata etc. so I appreciate that you go one by one. It brings clarity to it. Maybe at some point, we could move this code to a metadata component that would receive all content node's metadata in `props` and it could also accept a `prop` with a list of metadata keys in the desired order by which to filter metadata to be displayed (e.g. `[..., 'language', 'author', 'licenseHolder', ...]` for `CurrentlyViewedResourceMetadata` and  `..., ['language', 'accessibility, 'whatYouWillNeed', 'author', 'licenseHolder' ...]` for `BrowseResourceMetadata` ). However maybe it'd be better to look into it after we have the final version of the metadata structure?

**Truncation of description text**

I hope this new [(work in progress) truncator](https://github.com/learningequality/kolibri/pull/8464) could help. I think it should be possible to add view more/less functionality.

nucleogenesis committed:
Add thumbnails, make side panel X close-button z-index 24

nucleogenesis committed:
fixed import from relative to using api spec for i18n cxComponentTransltor

nucleogenesis committed:
use facilityConfig value for download button showing in learn

nucleogenesis committed:
genContentLink -> learn/utils

genContentLink was defined identically in ~half a dozen places

The ContentCardGroup* components and soon the SidePanel Browse metadata component use that function to take an ID and bool `isLeaf` then generate a route object with the proper name and params.

nucleogenesis committed:
Move the (Browse|CurrentlyViewed)ResourceMetadata classes to Learn

- Moves the files into learn/src/views - these components are unique to Learn's needs  and making them core would require more learn-specific components moving to core to avoid cross-plugin imports
- Provides test for new genContentLink util fn after the previous commit

nucleogenesis committed:
WIP Commit

MisRob commented on kolibri/core/assets/src/views/FullScreenSidePanel/SidePanelContentMetadata.vue:
```diff
@@ -0,0 +1,17 @@
+<template>
```
Is this component something to be implemented in the near future?

nucleogenesis replied:
Oops - no I think that I started an outline component but bailed on it when I realized I wanted to make two separate metadata components - will remove this one :)

MisRob commented on kolibri/core/assets/src/views/FullScreenSidePanel/index.vue:
```diff
@@ -127,15 +138,17 @@
-    z-index: 16;
-    width: 100vw;
+    bottom: 0;
+    // Must be <= 12 z-index so that KDropdownMenu shows over
```
Thank you for all explanations for z-index values

MisRob commented on kolibri/plugins/learn/assets/test/util/gen-content-link.spec.js:
```diff
@@ -0,0 +1,34 @@
+    expect(validateLinkObject(contentLink)).toBeTruthy();
+  });
+
+  it('returns an object with name pointing to PageName.TOPICS_TOPIC', () => {
```
returns an object with name pointing to PageName.TOPICS_**CONTENT**

nucleogenesis replied:
Great catch! `jest` caught my copy-pasting error in the code but not in the comments :-p

MisRob commented on kolibri/plugins/learn/assets/src/views/LearningActivityChip.vue:
```diff
@@ -0,0 +1,110 @@
+    [LearningActivities.EXPLORE]: 'interact',
+  };
+
+  /**
```
This doesn't describe the chip component precisely since it also displays a label in addition to an icon

MisRob commented on kolibri/plugins/learn/assets/src/views/LearningActivityChip.vue:
```diff
@@ -0,0 +1,110 @@
+<template>
+
+  <div class="activity-chip">
+    <KLabeledIcon :icon="icon" :label="coreString(kind)" class="activity-chip-label" />
```
Was there any reason for re-implementing the whole `LearningActivityIcon` logic from scratch in this component instead of using `LearningActivityIcon` in the icon slot of `KLabeledIcon`?

```
<KLabeledIcon :label="coreString(kind)" class="activity-chip-label" />
  <template #icon>
    <LearningActivityIcon :kind="kind" />
  </template>
</KLabeledIcon>
```

MisRob replied:
Regarding `:label="coreString(kind)"`, what is supposed to happen when `kind` is an array?

nucleogenesis replied:
I have not implemented the list version because - well I don't know what the data structure will look like when we get it.

IMO it should just always be an array and it should be called `activityTypes` but without the metadata in place I didn't see reason to put it until we know the shape (or potential shapes) of the incoming data.

nucleogenesis replied:
As for re-implementing, I think that I just hadn't really fully considered the existence of LearningActivytIcon - will import and use that here

MisRob replied:
> IMO it should just always be an array and it should be called activityTypes but without the metadata in place I didn't see reason to put it until we know the shape (or potential shapes) of the incoming data.

Good point. `activityTypes` will make much more sense than `kind` for `LearningActivityIcon` and similar components `props`. I don't recall why I allowed `String` type alongside `Array` but would say it wasn't anything important and might be caused by not knowing that we can have more activities for a content node in the beginning of development and adding support for that later. So I think it should be fine to simplify it to accept only `Array`. No need to do it in this PR for `LearningActivityIcon`

MisRob commented on kolibri/plugins/learn/assets/src/views/LearningActivityChip.vue:
```diff
@@ -0,0 +1,110 @@
+       * Icon is solid by default.
+       * Set to `true` to make it shaded.
+       */
+      shaded: {
```
From what I've seen in designs, I'd say that there's no need for the chip component to provide this settings in its API since chips are so small that they'll have `solid` learning activity icons:

![Screenshot from 2021-10-12 13-11-10](https://user-images.githubusercontent.com/13509191/136945456-f578b18d-9231-43f0-a00d-03f21f9be5b7.png)
![Screenshot from 2021-10-12 13-11-30](https://user-images.githubusercontent.com/13509191/136945460-eca35b2d-6e48-48cc-bc35-042aed05a271.png)

MisRob commented on kolibri/plugins/learn/assets/src/views/LearningActivityChip.vue:
```diff
@@ -0,0 +1,110 @@
+
+  .activity-chip {
+    display: inline-block;
+    padding: 12px; // Aligns with close button
```
Was this meant to refer to the close button in side panels? If yes, I'd suggest moving this positioning CSS to side panels CSS (maybe `chips` class?) because it's specific to the particular usage of this component in the side panel context and doesn't have much to do with the chip component itself.

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+    </div>
+
+    <div class="section">
+      <Thumbnail :thumbnailUrl="content.thumbnail" />
```
I made some updates to hybrid learning thumbnails and `Thumbnail` is now only a base component for other components (now I'm thinking it might make more sense to name it `BaseThumbnail`) Although this works, it won't display a learning activity icon when a thumbnail image is not available because this behavior is not part of `Thumbnail` implementation anymore. Could you please update thumbnails from `Thumbnail` to `ContentNodeThumbnail` so it works again as originally intended?

nucleogenesis replied:
The designs that I worked from didn't show the thumbnail so I just assumed that we wouldn't want the activity icon on the thumbnail in this context because the activity icon(s) are visible at the top of the side panel

MisRob replied:
The activity icon on the thumbnail only serves as a generic thumbnail placeholder when an image is not available so I don't consider having the activity icon somewhere close to the thumbnail a problem. There are already other places in the app when this is happening so if we really want to display the thumbnail here I'd suggest keeping it consistent. Here's one example [in designs](https://www.figma.com/file/uiZuCIqh2KYvBUfbCiJyyA/Hybrid-Learning?node-id=1915%3A82994) (note that later we made a decision to make activities icons on thumbnails a bit more colorful):

![Screenshot from 2021-10-13 13-11-40](https://user-images.githubusercontent.com/13509191/137122126-3bd822c4-67f2-4a10-829b-5d2f8279cef1.png)

Besides that, if we'd displayed only gray background without anything else, I'd say that it might be a bit confusing because giving a false impression that it's a placeholder for an image that will load soon but that won't actually happen.

nucleogenesis replied:
Ahh this was my mistake - I thought that it was a thumbnail that would overlay the text of the activity like "Read" overlaid on the thumbnail like is shown in the content browsing experience. This makes much more sense - will have a fix shortly.

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+      :primary="true"
+      @click="toggleShowMoreOrLess"
+    />
+    <!-- this v-else ensures spacing remains consistent without show more -->
```
Thank you, these template comments are helpful

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+          class="license-toggle"
+          @click="licenseDescriptionIsVisible = !licenseDescriptionIsVisible"
+        />
+        <div v-if="licenseDescriptionIsVisible" dir="auto" class="license-details">
```
Should automatic text directionality be used for all user-generated texts in Kolibri codebase or are there any other recommendations for `dir="auto"`? I can't understand why it's used only for the license description in this component (and often I am not sure when to add it to my code in Kolibri) so I'd be grateful for clarification.

nucleogenesis replied:
Honestly - I think I may have put it there while trying to debug an RTL issue and forgot to remove it - or I copy-pasted it. 

Perhaps @indirectlylit would have a better understanding of when/how we should use that dir attribute. 

In any case, this could go very well in our [KDS Inclusive Design ](https://design-system.learningequality.org/inclusive/) documentation... which is where I looked in case there was guidance there :-)

indirectlylit replied:
Good question!

* I have an open Kolibri issue with some guidance here: https://github.com/learningequality/kolibri/issues/3781
* Opened a new KDS issue here: https://github.com/learningequality/kolibri-design-system/issues/266

MisRob replied:
Thank you both

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+    },
+    computed: {
+      licenseShortName() {
+        return licenseShortName(this.content.license_name);
```
I'd suggest adding guards to check that `this.content` is truthy before accessing its attributes in this `computed` as well as in the following ones.

As long as `content` is indeed an `Object` as documented in `props`, accessing `this.content.<attribute>` should be fine, but I think that's rather information for developers and won't prevent from actually getting `undefined` or `null` through props so it could still fail here in such cases, I think?

rtibbles replied:
Yes, validation doesn't run in production, only in development.

MisRob replied:
Thank you

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+    padding: 10px;
+    font-weight: bold;
+    color: white;
+    background: #328168; // brand.secondary.v_600
```
Is there any reason for not using our brand color definition directly with computed class or style?

nucleogenesis replied:
I was tweaking my CSS in `devserver-hot` and didn't want to stop and do JS while I was in my styling groove :) - I'll fix this in my next pass.

MisRob commented on kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue:
```diff
@@ -0,0 +1,365 @@
+<template>
```
Most of the review comments for this component also apply to _CurrentlyViewedResourceMetadata.vue_

MisRob reviewed (approved):
Thank you @nucleogenesis.

Regarding code, I haven't noticed any blockers. It'd be good to have a look at the chip component and thumbnails.

Since Kolibri is still quite broken on this branch and side panels are work in progress, I didn't spend much time testing and only briefly checked panels toggling which seems to work well.

nucleogenesis committed:
remove empty old unused component

nucleogenesis committed:
fixed typo in gen-content-link test description

nucleogenesis committed:
fixed imports in LibraryPage post-rebase

nucleogenesis committed:
LearningActivityChips in sidepanel handle 1 or more learning_activities

nucleogenesis committed:
apply styles in parent component

nucleogenesis committed:
Conditionalize LearningActivityIcon, minor refactor

- Cleaned up some of the branching logic
- Icon computed property return null when given empty array
- Conditionalize KIcon on `if=icon` which may return null

nucleogenesis committed:
SidePanel metadata use ContentNodeThumbnail, remove errant dir=auto

nucleogenesis committed:
Use $themeBrand instead of the hex value I took from KDS :)

nucleogenesis committed:
guard against calls against undefined errors for license on content

nucleogenesis commented:
@MisRob if all checks pass, this should be ready for a final review

pcenov commented:
Hi @nucleogenesis while testing the [buildkite build](https://buildkite.com/learningequality/kolibri-python-package/builds/3890) I noticed that I cannot open the side panel from the Learn>Classes Tab and also that the Close icon can overlap other controls when scrolling down:
![2021-10-20_16-43-11](https://user-images.githubusercontent.com/79847249/138123980-1adb1f54-4de9-46fd-b9c1-e05330aa73a8.png)
![2021-10-20_16-48-57](https://user-images.githubusercontent.com/79847249/138124008-00aca54f-4511-4d87-9238-568aa4d91a71.png)
Other than that all looks fine to me. I'll report the above mentioned issues separately if you say that they are in scope/not already known to you.

nucleogenesis commented:
@pcenov thank you! I'm following up on related work and will be sure to hook that up
